File: web_iot/python/button_ws.py
import flask
import flask_socketio as io
import serial
import threading

dev_name = "/dev/tty.usbmodem14201"
speed = 9600
port = serial.Serial(dev_name, speed)
print("Serial device ready :", port.isOpen())

app = flask.Flask(__name__, static_folder="../static")
socketio = io.SocketIO(app, async_mode='threading')
isLive = True

@app.route('/')
def main():
    return flask.redirect('/static/button_ws.html')

@socketio.on('connect')
def connect():
    socketio.send({'data': 'welcome'})
    
def onChange():
    while isLive:
        val = port.readline().decode()
        val = eval(val)
        socketio.send( {'data': val })
        print('btn pressed "%s"'%val, type(val))


if __name__ == '__main__':
    try:
        t1 = threading.Thread(target=onChange)
        t1.start()
        socketio.run(app, host='0.0.0.0', port=8080)
    finally:    
        port.close()
        isLive = False

File: web_iot/nodeJs/button_ws.js
//npm install express serialport socket.io
const http = require('http');
const express = require('express');
const socketio = require('socket.io');
const SerialPort = require("serialport");
const readline = require('readline');

let app = express();
let server = http.createServer(app);
let io = socketio(server);

var dev_name = "/dev/tty.usbmodem14201"
var speed = 9600;
let port_r;
var port = new SerialPort(dev_name, {baudRate:speed}, (err)=>{
    if(err) console.log("serial port open err", err);
    else{
        console.log("serial port is ready.")
        port_r = readline.createInterface(port);
        port_r.on('line', (data)=>{
            console.log(data);
            io.emit('message', {data:data});
        });
    }
});

io.on('connect', (socket)=>{
    console.log("ws connect.");
    socket.emit('message', {'data': 'welcome'});

});
app.get('/', (req, res)=>{
    res.redirect('/button_ws.html');
});
app.use(express.static(__dirname + '/../static/'));

server.listen(8080, ()=>{
    console.log("server on 8080.");
});
